File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/DatabaseHolder.java
package com.raizlabs.android.dbflow.config;

import com.raizlabs.android.dbflow.converter.TypeConverter;
import com.raizlabs.android.dbflow.structure.Model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Description: The base interface for interacting with all of the database and top-level data that's shared
 * between them.
 */
public abstract class DatabaseHolder {
    protected final Map<Class<? extends Model>, DatabaseDefinition> databaseDefinitionMap = new HashMap<>();
    protected final Map<String, DatabaseDefinition> databaseNameMap = new HashMap<>();
    protected final Map<Class<?>, DatabaseDefinition> databaseClassLookupMap = new HashMap<>();
    protected final Map<Class<?>, TypeConverter> typeConverters = new HashMap<>();

    /**
     * @param clazz The model value class to get a {@link com.raizlabs.android.dbflow.converter.TypeConverter}
     * @return Type converter for the specified model value.
     */
    public TypeConverter getTypeConverterForClass(Class<?> clazz) {
        return typeConverters.get(clazz);
    }

    /**
     * @param table The model class
     * @return The database that the table belongs in
     */
    public DatabaseDefinition getDatabaseForTable(Class<? extends Model> table) {
        return databaseDefinitionMap.get(table);
    }

    public DatabaseDefinition getDatabase(Class<?> databaseClass) {
        return databaseClassLookupMap.get(databaseClass);
    }

    /**
     * @param databaseName The name of the database to retrieve
     * @return The database that has the specified name
     */
    public DatabaseDefinition getDatabase(String databaseName) {
        return databaseNameMap.get(databaseName);
    }

    /**
     * Helper method used to store a database for the specified table.
     *
     * @param table              The model table
     * @param databaseDefinition The database definition
     */
    public void putDatabaseForTable(Class<? extends Model> table, DatabaseDefinition databaseDefinition) {
        databaseDefinitionMap.put(table, databaseDefinition);
        databaseNameMap.put(databaseDefinition.getDatabaseName(), databaseDefinition);
        databaseClassLookupMap.put(databaseDefinition.getAssociatedDatabaseClassFile(), databaseDefinition);
    }

    public void reset() {
        databaseDefinitionMap.clear();
        databaseNameMap.clear();
        databaseClassLookupMap.clear();
        typeConverters.clear();
    }

    public List<DatabaseDefinition> getDatabaseDefinitions() {
        return new ArrayList<>(databaseNameMap.values());
    }
}

File: dbflow/src/main/kotlin/com/raizlabs/dbflow5/config/DatabaseHolder.kt
package com.raizlabs.dbflow5.config

import com.raizlabs.dbflow5.converter.TypeConverter

/**
 * Description: The base interface for interacting with all of the database and top-level data that's shared
 * between them.
 */
abstract class DatabaseHolder {

    val databaseDefinitionMap: MutableMap<Class<*>, DBFlowDatabase> = hashMapOf()
    val databaseNameMap: MutableMap<String, DBFlowDatabase> = hashMapOf()
    val databaseClassLookupMap: MutableMap<Class<*>, DBFlowDatabase> = hashMapOf()

    @JvmField
    val typeConverters: MutableMap<Class<*>, TypeConverter<*, *>> = hashMapOf()

    val databaseDefinitions: List<DBFlowDatabase>
        get() = databaseNameMap.values.toList()

    /**
     * @param clazz The model value class to get a [TypeConverter]
     * @return Type converter for the specified model value.
     */
    fun getTypeConverterForClass(clazz: Class<*>): TypeConverter<*, *>? = typeConverters[clazz]

    /**
     * @param table The model class
     * @return The database that the table belongs in
     */
    fun getDatabaseForTable(table: Class<*>): DBFlowDatabase? = databaseDefinitionMap[table]

    fun getDatabase(databaseClass: Class<*>): DBFlowDatabase? =
        databaseClassLookupMap[databaseClass]

    /**
     * @param databaseName The name of the database to retrieve
     * @return The database that has the specified name
     */
    fun getDatabase(databaseName: String): DBFlowDatabase? = databaseNameMap[databaseName]

    /**
     * Helper method used to store a database for the specified table.
     *
     * @param table              The model table
     * @param databaseDefinition The database definition
     */
    fun putDatabaseForTable(table: Class<*>, databaseDefinition: DBFlowDatabase) {
        databaseDefinitionMap.put(table, databaseDefinition)
        databaseNameMap.put(databaseDefinition.databaseName, databaseDefinition)
        databaseClassLookupMap.put(databaseDefinition.associatedDatabaseClassFile, databaseDefinition)
    }

    fun reset() {
        databaseDefinitionMap.clear()
        databaseNameMap.clear()
        databaseClassLookupMap.clear()
        typeConverters.clear()
    }
}
